plugins/quantum_k_nearest_neighbours/backend/qknns/basheer_hamming.py

```diff
@@ -0,0 +1,377 @@
+
+        def quantum_circuit():
+            # We do not use HX oracle XH here, since the oracle has a lot to uncompute and it only saves one Toffoli operation
+            # Set oracle wire to one
+            oracle_circuit()
+
+            # Set phase
+            qml.PauliZ((self.oracle_wire,))
+
+            # Uncompute oracle
+            qml.adjoint(oracle_circuit)()
+
+        return quantum_circuit
+
+    def idx_circuit(self):
+        for wire in self.idx_wires:
+            qml.Hadamard((wire,))
+
+    def zero_circuit(self):
+        # Check if idx_wires = |0>
+        for wire in self.idx_wires:
+            qml.PauliX((wire,))
+
+        qml.Hadamard((self.ancilla_wires[0],))
+        adaptive_ccnot(
+            self.idx_wires,
+            self.ancilla_wires[1:] + self.train_wires,
+            self.unclean_wires,
+            self.ancilla_wires[0],
+        )
+        qml.Hadamard((self.ancilla_wires[0],))
+
+        for wire in self.idx_wires:
+            qml.PauliX((wire,))
+
+    def get_better_training_point_idx(
+        self, x, distance_threshold, chosen_indices
+    ) -> Tuple[int, int]:
+        if distance_threshold < 0:
+            return None, None
+        state_circuit = self.idx_circuit
+        zero_circuit = self.zero_circuit
+
+        distance_threshold = min(distance_threshold, self.train_data.shape[0])
+        p = int(np.ceil(np.log2(self.train_data.shape[1])))
+        a = int(2**p - self.train_data.shape[1] + distance_threshold)
+        a = int_to_bitlist(a, p + 2)
+
+        oracle_phase_circuit = self.get_phase_oracle_circuit(x, a, chosen_indices)
+        oracle_one_circuit = self.get_oracle_wire_to_one_circuit(x, a, chosen_indices)
+        check_if_good_wire = self.oracle_wire
+        measure_wires = self.idx_wires + self.train_wires
+
+        result = exp_searching_amplitude_amplification(
+            state_circuit,
+            state_circuit,
+            zero_circuit,
+            oracle_phase_circuit,
+            self.backend,
+            oracle_one_circuit,
+            check_if_good_wire,
+            measure_wires,
+            exp_itr=self.exp_itr,
+        )
+
+        if result is None:
+            return None, None
+
+        idx_bits = result[: len(self.idx_wires)]
+        train_bits = result[len(self.idx_wires) :]
+        hamming_distance = len(train_bits) - np.array(train_bits).sum()
+        return int(bitlist_to_int(idx_bits) % self.num_train_data), hamming_distance
+
+    def label_point(self, x) -> int:
+        x = np.array(x, dtype=int)
+        # Init: First choose k random points, to be the current nearest neighbours
+        chosen_indices = np.random.choice(
+            range(self.num_train_data), self.k, replace=False
+        )
+
+        chosen_distances = np.array(
+            [calc_hamming_distance(x, self.train_data[idx]) for idx in chosen_indices]
+        )
+        converged = False
+
+        # Loop
+        while not converged:
```

Comment:
You can allow a bit of slack in the break condition to allow some errors from the quantum computation. If the computation exceeds the allowed iterations produce a meaningful error message and make sure it gets written to the task log (exception messages will end up in the task log, but you can write strings in there yourself too). Since current NISQ machines can be really noisy if the calibration is bad or the circuit is too large, it is good, that you include a break condition that always applies.

Reply:
I added an input parameter, with which the user can control the amount of slack. If it doesn't converge, I raise a UserWarning. I do not write any additional information into the task log, since the error message should suffice.